### PR TITLE
Speed up travis builds by caching webpack output

### LINKED ISSUE #2942: Cache yarn/webpack output for travis
Currently each container builds this separately, plus it takes quite a long time (~90s). To figure out how to cache this if there are no javascript changes.

### PR DESCRIPTION
**Pull request by fonglh**
Cache the contents of the `public/webpack` directory.

Run `yarn build:production` if there is a `git diff` in the `client` folder between the PR and master, or if `public/webpack/manifest.json` does not exist.

The cache *could* be wrong if a change in the `client` folder is added to a PR then subsequently removed such that there is no difference with the master branch.

Fixes #2942.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -45,7 +46,7 @@ before_install:
 before_script:
   - psql -c 'create database coursemology_test;' -U postgres
   - psql -c 'create database coursemology_test2;' -U postgres
-  - cd client && yarn build:production && cd -
+  - 'if ! git diff --exit-code master client || ! cat $TRAVIS_BUILD_DIR/public/webpack/manifest.json; then cd client && yarn build:production && cd -; fi'
```
Comment (weiqingtoh) [inline .travis.yml]: Should we explain what this command is trying to do?

**Comment by weiqingtoh** [comment]
> The cache _could_ be wrong if a change in the `client` folder is added to a PR then subsequently removed such that there is no difference with the master branch.

Good point - is there a way for us to detect such changes, otherwise there might be issues because we will be unable to invalidate the cache >_<

**Comment by weiqingtoh** [comment]
There are webpack plugins that help to cache and build up speed times - I tested one that does reduce build times by around 30-40% -> https://github.com/mzgoddard/hard-source-webpack-plugin.

Perhaps you can look into these solutions instead?

**Comment by fonglh** [comment]
> The other comment is also that there are javascript files in the app/assets/javascript folder, I believe they are also processed through webpack. If you find another solution, do include that folder as well.

The files in `app/assets/javascript` are processed by the `rake assets:precompile` command. The output of `yarn build:production` does not change when a file in `app/assets/javascript` is changed.

It looks like there are 2 different invocations of webpack, 1 for Rails and 1 for the javascript in the `client` folder.

**Comment by fonglh** [comment]
Trying out hard source webpack plugin.

Meanwhile, will put the manual git diff commands here for future reference.

```
# Run `yarn build:production` command when the client folder in this commit is different from the master branch,
  # or if the public/webpack/manifest.json file is missing (implies no cached js files).
  #- 'if ! git diff --exit-code master client || ! cat $TRAVIS_BUILD_DIR/public/webpack/manifest.json; then cd client && yarn build:production && cd -; fi'
```

Other suggestions include using tar to generate a checksum on the whole folder and storing that instead. See https://unix.stackexchange.com/questions/35832/how-do-i-get-the-md5-sum-of-a-directorys-contents-as-one-sum for ideas.

**Comment by fonglh** [comment]
Can't tell whether this works yet because travis does not seem to use the updated cache on additional commits. Perhaps we can merge first and see if it makes a difference when there are other PRs?

**Comment by fonglh** [comment]
Ok can, thanks!

**Pull request merged**